python/taulabs/telemetry.py

```diff
@@ -177,28 +199,62 @@ def serviceConnection(self, timeout=None):
         else:
             finishTime = None
 
-        data = self.__receive(finishTime)
-        self.__handleFrames(data)
+        data = self._receive(finishTime)
+        self._handleFrames(data)
 
-    def __send(self, msg):
-        """ Send a string out the TCP socket """
+    @abstractmethod
+    def _receive(self, finishTime):
+        return
 
-        self.send_buf += msg
+    # No implementation required, so not abstract
+    def _send(self, msg):
+        return
+
+    @abstractmethod
+    def _done(self):
+        return True
+
+class FDTelemetry(TelemetryBase):
+    # expects fd is set nonblocking
+    # intended for bidirectional comms.
+    def __init__(self, fd, *args, **kwargs):
+        TelemetryBase.__init__(self, doHandshaking=True,
+                gcsTimestamps=False,  *args, **kwargs)
 
-        self.__select(0)
+        self.recv_buf = ''
+        self.send_buf = ''
+
+        self.fd = fd
+
+    def _receive(self, finishTime):
+        """ Fetch available data from TCP socket """
```

Comment:
This wouldn't have to be strictly a TCP socket now that you've abstracted out the FDTelemetry class.